Instant verification support for L2OutputOracle and Bridge

ironbeer opened the pull request.

ironbeer committed:
Instant verification support for L2OutputOracle and Bridge

ironbeer committed:
Immediate withdraw only from the allowed relayer

ironbeer committed:
Add batch proven method

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,173 @@
+/// @notice The OasysPortal is TODO
+contract OasysPortal is OptimismPortal {
+    /// @notice Message relayer to allowed immediate withdraw.
+    address public relayer;
```
`messageRelayer`に変更願いします🙏
relayerは一般的な名前なのでぱっと見、何かわからないので。

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    /// @notice Proves withdrawal transactions.
+    function proveWithdrawalTransaction(
```
必須な処理しか追加したくない思いがあります。
Oasysの拡張は最小限にしたいので。
なので、機能を増やすのは慎重に考えたいです。
コードも長くなったり、インターフェースが変わったらここも変更しなくちゃいけなかったりも、しますし。

これはmulticallで代替できるので、オプショナルな実装だと思っています。
（少しガス代が安くなるメリットはありますが、微々たるものだと思っています。

ironbeer replied:
>これはmulticallで代替できるので、オプショナルな実装だと思っています。

Multicallを通すとmsg.senderがmulticallのアドレスになってしまうんですよね。

tak1827 replied:
確かに！
proveの方はmsg.sender依存の処理はない？と思うので消すとして、
finalizeの方は残しておいてください。なぜ、必要なのかの理由をコメントで添えて

ironbeer replied:
@tak1827 このような処理に変えてみましたがどうでしょうか？　https://github.com/oasysgames/oasys-opstack/pull/61/files#diff-90ee50470dc8112c8ab0752c8905531b3e0d40d90fae9b20863418c002c41addR64-R75

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    /// @notice Finalizes a withdrawal transaction.
+    function finalizeWithdrawalTransaction(Types.WithdrawalTransaction calldata _tx) public override whenNotPaused {
```
オリジナルをそもまま使いたいです。

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,173 @@
+        // finalized. This waiting period can elapse in parallel with the waiting period for the
+        // output the withdrawal was proven against. In effect, this means that the minimum
+        // withdrawal time is proposal submission time + finalization period.
+        if (msg.sender != relayer) {
```
このチェックを`_isFinalizationPeriodElapsed`側で行いたいです。
関数の意味するところが変わるので、コメントで補足しつつ

ironbeer replied:
`isOutputFinalized()`というview関数からも参照されている内部関数ですのでmsg.senderを参照する処理を入れるのは微妙ではないでしょうか？
`_isAllowFinalizeWithdrawalTransaction(address sender, uint256 timestamp)`の様な別関数ではNGでしょうか？

tak1827 replied:
> view関数からも参照されている内部関数ですのでmsg.senderを参...

特に違和感はないです。設計として意識したこともないです。
しかし、よく考えてみると、erc20とかbalanceOfでアドレスを指定していますね。

> _isAllowFinalizeWithdrawalTransaction(address sender, uint256 timestamp)の様な別関数...

finalizeWithdrawalTransactionをコピペしつつ、置き換えるということですよね？
オリジナルに変更があったとっきに、見比べるの大変なので、なるべくコピペしたくはないです。。。
機能に変わりはないので、最終的にどうするかは任せします。

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    /// @notice Finalizes withdrawal transactions.
+    function finalizeWithdrawalTransaction(Types.WithdrawalTransaction[] calldata txs) external {
```
これも、オプショナルだと思っています。

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,173 @@
+    /// @param _timestamp Timestamp to check.
+    /// @return Whether or not the finalization period has elapsed.
+    function _isFinalizationPeriodElapsed(uint256 _timestamp) internal view override returns (bool) {
+        uint256 verified = IOasysL2OutputOracle(address(L2_ORACLE)).verifiedL1Timestamp();
```
気の利いたコメントが欲しいです。

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/rollup/OasysL2OutputOracle.sol:
```diff
@@ -38,66 +35,100 @@ contract OasysL2OutputOracle is L2OutputOracle {
 
-        require(_batchHeader.batchIndex == nextIndex, "Invalid batch index.");
+    /// @inheritdoc IOasysL2OutputOracle
+    function verifiedBlockNumber() external view returns (uint256) {
```
`nextBlockNumber`の追となる実装でしょうか？
無くても良いんじゃないかと思います。
Static Block Timeの時しか正しい値は返しませんし、Dynamicの時は混乱すると思うので。

ironbeer replied:
ロールアップとベリファイは `SUBMISSION_INTERVAL`で指定されたブロック数毎に行われますのでブロック時間は無関係ですね。
https://github.com/oasysgames/oasys-opstack/blob/d95e776da89e73f077d49d18ae3c6574aef043e8/op-proposer/proposer/driver.go#L162

ただこちらのメソッドはfinalize実行者が`if l2oo.verifiedBlockNumber(withdrawTx.blockNumber)`の様に使用する事を想定していたのですが「message relayerアドレスかどうか？」を見る実装に変わったことにより不要になりましたので削除します。

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/rollup/OasysL2OutputOracle.sol:
```diff
@@ -38,66 +35,100 @@ contract OasysL2OutputOracle is L2OutputOracle {
-        super.proposeL2Output(_outputRoot, _l2BlockNumber, _l1BlockHash, _l1BlockNumber);
+        require(_isValidL2Output(l2OutputIndex, l2Output), "OasysL2OutputOracle: invalid output root");
+
+        _deleteL2Outputs(l2OutputIndex);
```
この機能を引き継ぐかどうか、議論の余地がありそうです。
instant verifierの機能は「L2->L1ブリッジの高速化のみ」としてもいい気がします。

ironbeer replied:
verifierはapprove署名のみ作成し、問題あるロールアップがされた場合はverse側の責任でdeleteさせるイメージでしょうか？

ironbeer committed:
- Delete batch methods
- Rename `relayer` => `messageRelayer`

ironbeer committed:
Delete `verifiedBlockNumber` method

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/rollup/OasysL2OutputOracle.sol:
```diff
@@ -38,66 +35,95 @@ contract OasysL2OutputOracle is L2OutputOracle {
+    }
 
-        nextIndex += SUBMISSION_INTERVAL;
+    function _deleteL2Outputs(uint256 l2OutputIndex) internal {
```
場合によっては`nextVerifyIndex`も減算してください。
L1のReorgにともない、L2もReorgしたときとか。

ironbeer replied:
nextVerifyIndex未満のl2OutputIndex(つまりverify済み)は削除禁止しているので減算処理は不要ですね。
https://github.com/oasysgames/oasys-opstack/pull/61/files/d87f3bbe29422b8e9abd35196d7b08ec2fb241d7#diff-f2c29189a9192bd0269dce91818e2a1d8b29ac320fa24de1cddf8651bb8cf64aR121

L1がReorgされた場合はnextVerifyIndexも巻き戻るので大丈夫かと思います。

ironbeer committed:
Rename L2OutputOracle build contract

ironbeer committed:
Add batch withdraw method

ironbeer committed:
- Removed the batch proof method.
- Instead of overriding the Finalize method, it now checks for a message relayer within _isFinalizationPeriodElapsed.

ironbeer committed:
Fixed undeclared identifier error

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/lib/OasysPredeploys.sol:
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: MIT
```
`PredeployAddresses`があるのに、あえて用意したのは、0.8.15でBuildしたいから、でしょうか？
それとも、SCC_VERIFIERを入れ替えるためでしょうか？

ironbeer replied:
削除しました。ファイル配置の整理をしようとしていたのでその名残です。

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,75 @@
+
+/// @custom:proxied
+/// @title OasysPortal
+/// @notice The OasysPortal is TODO
```
このTODOは消し忘れ、でしょうか？

ironbeer replied:
description修正しました。

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,75 @@
+/// @notice The OasysPortal is TODO
+contract OasysPortal is OptimismPortal {
+    /// @notice Message relayer to allowed immediate withdraw.
+    address public messageRelayer;
```
OptimismPortalがアップデートされたときのslotと競合するので、この前に、__gapをおいてください。
messageRelayerの前か、OptimismPortalのpauseの後に。
```
bool public paused;

/// @notice Reserve extra slots in the storage layout for future upgrades.
///         A gap size of 47 was chosen here, the first(l2Sender) and second(paused) slots are used
///         child contract would be a multiple of 50.
uint256[47] private __gap;
```

ironbeer replied:
OptimismPortal側に追加しました。
https://github.com/oasysgames/oasys-opstack/pull/61/files#diff-f5f7dc90c748b7d98110fa1cc80f3bedd5aa9f299b691c51b9bb473a20b526caR71

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,75 @@
+    ///         If the zero address is set, no immediate relay of withdrawal messages.
+    function setMessageRelayer(address newRelayer) external {
+        require(msg.sender == GUARDIAN, "OasysPortal: only guardian can set a new message relayer");
+        require(newRelayer != messageRelayer, "OasysPortal: already set");
```
updateできるようにしたいです。

そして、messgeRelayerを廃止する場合を考慮して、あえて、ゼロ値でないことにチェックはしない方が良いと思われます。
というような旨をコメントで残しておいてください。Audit対策で。

あと、private-opstack用のBuild時のステップに組み入れて欲しいです。

ironbeer replied:
> あと、private-opstack用のBuild時のステップに組み入れて欲しいです

追加しました。
https://github.com/oasysgames/oasys-opstack/pull/61/files#diff-cf02e2390824288c43417a877c05aac776d0bebcd1fd653c4e96ba8924526bfdR493-R516

private-opstackでの設定手順も追加しました。
https://github.com/oasysgames/oasys-private-opstack/commit/8057812622a6a8c4273a32f316bb4add32a1a36

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,75 @@
+    { }
+
+    /// @inheritdoc OptimismPortal
+    function initialize(bool _paused) public override {
```
この処理を敢えて書いたのは、コンパイラにoverrideを強制されたのでしょうか？

ironbeer replied:
はい、BuildOasysPortalがコンパイルエラーになります。
```
Compiler run failed:
Error (9582): Member "initialize" not found or not visible after argument-dependent lookup in type(contract OasysPortal).
  --> src/oasys/L1/build/BuildOasysPortal.sol:38:31:
   |
38 |         return abi.encodeCall(OasysPortal.initialize, (_paused));
   |
```

`abi.encodeCall(OptimismPortal.initialize, (_paused))`と書けばエラーは消えますが一長一短ありますね。

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/messaging/OasysPortal.sol:
```diff
@@ -0,0 +1,75 @@
+    function _isFinalizationPeriodElapsed(uint256 _timestamp) internal view override returns (bool) {
+        if (messageRelayer != address(0) && msg.sender == messageRelayer) {
+            // Note: If called with `eth_call`, this code block
+            //       will not pass because `msg.sender` is fixed at zero address.
```
fromを指定しても、zero addressなのでしょうか？
https://www.quicknode.com/docs/ethereum/eth_call

ironbeer replied:
from指定可能でしたのでコメント削除しました。

tak1827 commented on packages/contracts-bedrock/src/oasys/L1/rollup/OasysL2OutputOracle.sol:
```diff
@@ -1,22 +1,19 @@
-    event StateBatchFailed(uint256 indexed _batchIndex, bytes32 _batchRoot);
+contract OasysL2OutputOracle is IOasysL2OutputOracle, L2OutputOracle {
+    /// @notice Next L2Output index to verify.
+    uint256 public nextVerifyIndex;
```
同様に__gapを入れて欲しいです。

ironbeer replied:
こちらも、L2OutputOracle側に追加しました。
https://github.com/oasysgames/oasys-opstack/pull/61/files#diff-f92c9308509097bf11b443e68883b1ed5a542f9d306988cdec0431831d9f1f5aR49

ironbeer committed:
Removed unnecessary file

ironbeer committed:
Add contract description

ironbeer committed:
Add reserved slots

ironbeer committed:
Removed incorrect comment

ironbeer committed:
Merge branch 'develop' into feat/instant-verify

ironbeer committed:
Add foundry script to set message relayer in the OasysPortal

tak1827 commented on packages/contracts-bedrock/src/L1/L2OutputOracle.sol:
```diff
@@ -45,6 +45,9 @@ contract L2OutputOracle is Initializable, ISemver {
     Types.OutputProposal[] internal l2Outputs;
 
+    /// @notice Reserve extra slots (to a total of 50) in the storage layout for future upgrades.
+    uint256[50] private __gap;
```
50ではなく、45です。
継承しているInitalizableで2slot(_initialized, _initializing)使っていて、
L2OutputOracleせ、slot(startingBlockNumber, startingTimestamp)使っています。
Opstackでは継承先がちょうど50の倍数で始まるように設計さているので、
49-2-2=45です。

ironbeer replied:
予約スロット数調整しました。
```
# Initializable(1スロット消費)
    uint8 private _initialized;  = Slot #0
    bool private _initializing;  = Slot #0 (32バイト以下の連続する変数はパックされる)

# L2OutputOracle(3スロット消費)
    uint256 public startingBlockNumber;         = Slot #1
    uint256 public startingTimestamp;           = Slot #2
    Types.OutputProposal[] internal l2Outputs;  = Slot #3
```

tak1827 commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -67,6 +67,9 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
     bool public paused;
 
+    /// @notice Reserve extra slots (to a total of 50) in the storage layout for future upgrades.
+    uint256[50] private __gap;
```
ここも同様に計算する必要があります。

ironbeer replied:
こちらも調整しました。ResourceMeteringが__gapを確保してくれているので継承元を考慮する必要はありません。

ironbeer committed:
Modify reserved slots

ironbeer committed:
Modify reserved slots
